add cabal-docspec to flake dependencies and include examples for most symbols

MMesch opened the pull request:
This adds `cabal-docspec` to the flake dependencies to start addressing https://github.com/VinylRecords/Vinyl/blob/master/vinyl.cabal#L96.

It's currently taken from the Nix user repo because it is not yet on hackage or nixpkgs.

Run with

```
nix develop
cabal-docspec
```

This is the current output summary:

```
[   0.60765] doctest.summary: 
Total:        77; Tried:   36; Skipped:   41; Success:   30; Errors:    6; Failures    0
Examples:     75; Tried:   34; Skipped:   41; Success:   28; Errors:    6; Failures    0
Setup:         2; Tried:    2; Skipped:    0; Success:    2; Errors:    0; Failures    0
```

I plan fixing the errors in subsequent PRs, but I thought it'd be nice to have cabal-docspec in already before because there seems to be no possibility to run doctests at the moment whatsoever. And also to keep PR size small.

MMesch committed:
add cabal-docspec to flake dependencies

Currently from the Nix user repo because it is not yet on hackage or
nixpkgs

MMesch committed:
fix CoRec test for match

MMesch committed:
fix Tutorial doctest

Unfortunately this requires introducing singletons as dependencies.
I don't know how to best solve this.

MMesch committed:
fix Curry doctests

MMesch committed:
align stack LTS with flake's ghc8107

MMesch commented:
I am working now on this branch but maybe it'd be better to split the PR up. @acowley let me know what you would prefer. Also, I really don't know what the best solution is to make the Tutorial pass doctests without adding a global dependency on singletons which is a bit annoying.

acowley commented:
How would you split it up?

MMesch commented:
There are some simple changes such as fixing or adding doctests. And there are some like the singletons version bounds that may be more complex.

I pushed a new version with open version bounds but I'm unsure how to sustainably resolve this issue because it's a pity to add a dependency on the main library just for the tutorial. Maybe best would be to keep the tutorial just in tests and then link to it from the haddock page?

MMesch committed:
Add examples for Rec

MMesch committed:
add example for rappend

MMesch committed:
add example for <+>

MMesch committed:
add haskell language server to devShell

MMesch committed:
Add example to rcombine

acowley commented:
That sounds reasonable regarding `singletons`. I agree that we should avoid making it a dependency of the library if possible.

MMesch committed:
Add example for rmap

MMesch committed:
add example for <<$>>

MMesch committed:
add example for <<&>>

MMesch committed:
add example for rapply

MMesch committed:
add example for <<*>>

MMesch committed:
add example for rpure

MMesch committed:
add example for rtraverse

MMesch committed:
add example for rtraverseIn

MMesch committed:
add example for rsequenceIn

MMesch committed:
add another example for rtraverse

MMesch committed:
add another example for rtraverse

MMesch committed:
add example for rzipWith

MMesch committed:
add example for rfoldMap

MMesch committed:
add example for Dict

MMesch committed:
add example for reifyConstraint

MMesch committed:
add example for rpureConstrained

MMesch committed:
add example for DictOnly

MMesch committed:
add example for withPairedDict

MMesch committed:
add example for data Nat

MMesch committed:
add example for natToInt

MMesch committed:
add example for indexWitnesses

MMesch committed:
add example for Fst

MMesch committed:
add example for Snd

MMesch committed:
add example for RLength

MMesch committed:
add example for RIndex

MMesch committed:
add example for RImage

MMesch committed:
add example for RDelete

MMesch committed:
add example for RecAll

MMesch committed:
add example for (++)

MMesch committed:
add example for AllConstrained

MMesch committed:
illustrate issue with AllSatisfied class

MMesch committed:
add AllSatisfied', alternative for AllSatisfied

The alternative is computed as type family not as class. The latter
doesn't seem to work.

MMesch committed:
replace AllSatisfied with AllSatisfied'

MMesch committed:
add example for AllAllSat

MMesch committed:
add example for rpureConstraints

MMesch committed:
example for show instance

MMesch committed:
example for Semigroup instance

MMesch committed:
example for Monoid instance

MMesch committed:
add example for Eq instance

MMesch committed:
add example for Ord instance

MMesch committed:
add example for Head

MMesch committed:
add example for Tail

MMesch committed:
remove doctest dependency

MMesch committed:
add cabal-plan to dev tools

MMesch commented:
@acowley I think I managed to run all examples, including the existing Tutorial through cabal-docspec without adding singletons as a library dependency (it is added through a dummy doctests target and then exposed through an extra field).

I think we could try activating it in CI but that's a bit difficult for me to do without access to it. In theory you should be able to run it with `cabal-docspec` in the base folder after building the code with `cabal build all --enable-tests`.

acowley commented on flake.nix:
```diff
@@ -3,12 +3,16 @@
-
+      nurpkgs = import nur {
+          nurpkgs = nixpkgs.legacyPackages.x86_64-linux;
+          pkgs = nixpkgs.legacyPackages.x86_64-linux;
```
This doesn't evaluate on `aarch64` systems, and the `cabal-docspec` in `nur` seems be [downloading a binary](https://github.com/nix-community/nur-combined/blob/1207aa33351d6fd3a68de49326431effedbf3e4a/repos/amesgen/pkgs/default.nix#L93).

acowley replied:
Can we build `cabal-docspec` as a normal Haskell package in our flake?

MMesch replied:
That may be possible. Not sure when I'll have the time to play around with that but it seems the easiest path forward. Too bad it isn't on hackage.

MMesch replied:
There is also this https://github.com/phadej/cabal-extras/issues/134

MMesch commented:
@acowley , I investigated this a bit and unfortunately it seems more complex to get a cabal-docspec binary for Mac and Windows than what I have time for right now.

Wouldn't it be sufficient to just run the doctests on the Linux CI? As opposed to the build, I don't think testing the docs cross platform adds much extra value. And for development on mac, maybe the cabal-docspec Docker container works until there is a proper binary available?

acowley commented:
Yes, running only in CI is definitely an option. I was going to take a look at a from source build of `cabal-docspec`, did anything in particular not work out when you looked at it or was it just a big pain?

MMesch commented:
Nothing in particular except for various general comments in the cabal-extras repo on the build process that _may_ indicate complexity. But it's probably more because I'm not super experienced with the Nix Haskell infra, I fear I can't really debug any Macos or Windows builds, and I don't have enough headspace at the moment to dive into that 🙄. I want to mostly add documentation to vinyl and frames and maybe make some smaller improvements since I tried using those more actively recently. 

I also wonder what the proper way forward is so that we don't duplicate effort that should find it's place upstream. But considering all options (eg hackage doesn't seem to be in sight), I think writing a cross platform flake for cabal-extras seems like the most reasonable thing to do.

acowley commented:
Wow, `cabal-extras` is pretty deeply idiosyncratic. I'm not sure if it's worth trying to make its individual parts look more like normal packages. Definitely open to the CI-only solution.

MMesch commented:
@acowley , sorry this ready except for running it cleanly in the CI. I think it's a useful addition but I'm a bit blocked on my own. Do you think we can somehow finish it together?

MMesch commented:
ping @acowley
